### PR TITLE
fix(streaming): less exchange permits (and chunk_size)

### PR DESCRIPTION
**Pull request by fuyufjh**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The motivation is to fix the OOM failure of TPC-H Q2 (#10330), and also improve the streaming latency for any streaming queries.

Check https://github.com/risingwavelabs/risingwave/issues/10330#issuecomment-1594151474 for more context. In short, the root cause is that the buffered message (StreamChunks) amid the graph used too much memory and caused OOM at some point.

As expected, this will decrease the streaming performance a little bit, but I think it worths.

#### [Performance Comparison](http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/1956-nexmark-blackhole-risingwave-vs-flink?flink_metrics=avg-job-throughput-per-second&risingwave_metrics=avg-source-output-rows-per-second&flink_version=v1.16.0&risingwave_tag=nightly-20230614&flink_labels=v1.16.0-20230530&rw_labels=compute-compactor-anti-affinity%2Cchunk-size-256-barrier-1)

image

#### [TPC-H Q2 Passed](https://buildkite.com/risingwave-test/tpch-benchmark/builds/858#0188c2a7-20b2-4d3a-9287-89e37fae0bc9)

By the way, I have also tried keeping the chunk_size but only decreasing `concurrent_barrier_limit` to 1, but [the test failed](https://buildkite.com/risingwave-test/tpch-benchmark/builds/859#0188c2d1-6a86-4545-b6d1-a45b6508cd42).

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/risedevtool/src/task/compactor_service.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ impl Task for CompactorService {
             // See https://linux.die.net/man/3/jemalloc for the descriptions of profiling options
             cmd.env(
                 "MALLOC_CONF",
-                "prof:true,lg_prof_interval:38,lg_prof_sample:19,prof_prefix:compactor",
+                "prof:true,lg_prof_interval:34,lg_prof_sample:19,prof_prefix:compactor",
```
Comment (fuyufjh) [inline src/risedevtool/src/task/compactor_service.rs:102]: Intended to make future profiling easier.

**Comment by BugenZhao** [comment]
Have we tested only decreasing the "total permits"? I guess making the chunk size to be 256 hurts the most due to negative effects on vectorization and more overhead from dynamic dispatches.

BTW, it's interesting to find that this configuration influences stateless queries most, while there's no other memory usage for these queries so the risk of OOM is also lower. Perhaps we really need to introduce some dynamic strategies for the permit values. 🥵

**Review by lmatz — APPROVED** [review]

**Comment by lmatz** [comment]
> perhaps we really need to introduce some dynamic strategies for the permit values.

Agree, that can be a long-term solution in the future, e.g. decided by the front end according to the largest number of exchanges among all the paths from source to sink of a job

**Comment by fuyufjh** [comment]
> Have we tested only decreasing the "total permits"?

Good idea. Just ran a benchmark: https://buildkite.com/risingwave-test/tpch-benchmark/builds/862

Previously I thought that chunk_size also determines the memory usage, more or less. But you are right, if that can solve the problem, I prefer not to touch the chunk_size

**Comment by BugenZhao** [comment]
> Have we tested only decreasing the "total permits"?

[Updated comparisons](http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/4157-nexmark-blackhole-risingwave-executions-by-tag-labels?metrics_name=avg-source-output-rows-per-second&tag=nightly-20230614&labels=compute-compactor-anti-affinity%2Cchunk-size-256-barrier-1%2Ceric-reduce-exchange-permits-only-2048-nexmark): 


Screenshot 2023-06-21 at 2 41 51 PM

**Comment by fuyufjh** [comment]
> > Have we tested only decreasing the "total permits"?
> 
> [Updated comparisons](http://risingwave-perf-test-dashboard-metabase.us-west-2.elasticbeanstalk.com/question/4157-nexmark-blackhole-risingwave-executions-by-tag-labels?metrics_name=avg-source-output-rows-per-second&tag=nightly-20230614&labels=compute-compactor-anti-affinity%2Cchunk-size-256-barrier-1%2Ceric-reduce-exchange-permits-only-2048-nexmark):
> 
> Screenshot 2023-06-21 at 2 41 51 PM
> TPC-H Q2 succeeded under this configuration as well. I'm unsure if we need to adjust the numbers further for other more complex queries...

Unfortunately, OOM still happened... (but the test didn't fail, is it expected? cc. @cyliu0)

https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-test-dashboard?orgId=1&from=1687246527000&to=1687247309000&var-namespace=eric-tpch-q2-concurrent-barrier-1

image

**Comment by cyliu0** [comment]
> but the test didn't fail, is it expected?

Actually, it's failed all the time. It appeared in every day's brief like this. And this is quite an old known issue because of the OOM. I suppose everyone knows this. :(
```
Failed Execution List:
- tpch-q2-bs-medium-1cn: 4787
```

**Comment by BugenZhao** [comment]
> Unfortunately, OOM still happened...

IMO, 1024 rows in a chunk and 2 chunks in a buffer are really reasonable for a pipelined and vectorized system. 🥵 If this is still problematic, I would suggest decreasing the parallelism, upgrading the machine, or introducing dynamically allocated buffers. 😕

**Comment by fuyufjh** [comment]
> > Unfortunately, OOM still happened...
> 
> IMO, 1024 rows in a chunk and 2 chunks in a buffer are really reasonable for a pipelined and vectorized system. 🥵 If this is still problematic, I would suggest decreasing the parallelism, upgrading the machine, or introducing dynamically allocated buffers. 😕

Yeah, 1024 looks like a typical setting for batch vectorized execution engine. While streaming is somehow special because it also has to trade off between latency and throughput. Larger chunk also increases the latency, because `total latency = streaming processing time + time of waiting for chunk_size rows`. Does this make sense to you? 

"Decreasing the parallelism" also looks good. Q2 runs stably under `parallelism=4` with same machine specs. But we don't have any better idea yet, right...

**Comment by BugenZhao** [comment]
> Larger chunk also increases the latency, because basically `total latency = streaming processing time + time of waiting for chunk_size rows`. Does this make sense to you?

I'm also considering this... Using a large chunk size causes operators to be more likely to buffer data in the memory and yield them all at once on the barrier, which also increases the total latency since this procedure is not pipelined.

FYI, we use [`ready_chunks`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.ready_chunks) when doing data buffering in the connector, which seems to be a good balance between batching and latency. Perhaps we should apply a similar idea to other stateful executors as well.

**Comment by fuyufjh** [comment]
@BugenZhao So do you think it's okay to merge this PR? If so, please just click "Merge when ready" to avoid ping-pong

**Review by BugenZhao — APPROVED** [review]

**Comment by BugenZhao** [comment]
I'll help to fix the unit tests.

**Pull request merged**